crates/ide/src/ide/inlay_hints.rs

```diff
@@ -0,0 +1,236 @@
+use crate::{DefDatabase, FileId};
+use itertools::Itertools;
+use smol_str::ToSmolStr;
+use std::fmt::Display;
+use std::num::NonZero;
+use syntax::ast::{self, AstNode};
+use syntax::{SyntaxKind, SyntaxToken, TextRange};
+
+#[derive(Debug, Clone)]
+pub enum InlayHintKind {
+    LetBindingEnd(String),
+    AttrsetBindingEnd(String),
+    RecAttrsetBindingEnd(String),
+}
+
+impl Display for InlayHintKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::LetBindingEnd(s) => write!(f, "= let {s}"),
+            Self::AttrsetBindingEnd(s) => write!(f, "= {s}"),
+            Self::RecAttrsetBindingEnd(s) => write!(f, "= rec {s}"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct InlayHintResult {
+    pub range: TextRange,
+    pub kind: InlayHintKind,
+}
+
+#[derive(Debug, Clone)]
+pub struct InlayHintsConfig {
+    pub binding_end_hints_min_lines: Option<NonZero<usize>>,
+}
+
+pub(crate) fn inlay_hints(
+    db: &dyn DefDatabase,
+    file: FileId,
+    range: Option<TextRange>,
+    config: InlayHintsConfig,
+) -> Vec<InlayHintResult> {
+    let root_node = db.parse(file).syntax_node();
+
+    let (first_tok, end_pos) = match range {
+        None => (root_node.first_token(), u32::MAX.into()),
+        Some(range) => (
+            root_node.token_at_offset(range.start()).right_biased(),
+            range.end(),
+        ),
+    };
+
+    let binding_end_hints_min_lines = config
+        .binding_end_hints_min_lines
+        .unwrap_or(NonZero::new(25).expect("25 is nonzero"));
```

Comment:
I'm expecting `None` means to disable this hint? We may add more inlay hints in the future so this function still needs to be called.

Most default configuration values are set in outermost [`Config`](https://github.com/oxalica/nil/blob/9e4cccb088440c20703d62db9de8d5ae06d4a449/crates/nil/src/config.rs#L66).